SwiftIsland/Timeline/TimelineViewController.swift

```diff
@@ -151,64 +158,94 @@ private extension TimelineViewController {
   }
 }
 
+// MARK: UITableViewDelegate
 extension TimelineViewController: UITableViewDelegate {
 
   func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
-    let activity = activities[indexPath.section][indexPath.row]
+    let rowCount = self.tableView(tableView, numberOfRowsInSection: indexPath.section)
+    guard indexPath.row > 0, indexPath.row < rowCount - 1 else {
+      return
+    }
+    let adjustedIndexPath = IndexPath(item: indexPath.item - 1, section: indexPath.section)
+    let activity = activities[adjustedIndexPath.section][adjustedIndexPath.row]
     showActivity(activity: activity)
   }
 }
 
+// MARK: UITableViewDataSource
 extension TimelineViewController: UITableViewDataSource {
 
   func numberOfSections(in tableView: UITableView) -> Int {
     return schedule.count
   }
 
   func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-    return activities[section].count
+    return activities[section].count + 2
   }
 
   func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+
+    let rowCount = self.tableView(tableView, numberOfRowsInSection: indexPath.section)
+    // We avoid using a section header and footer view, opting for actual cells instead.
+    // This way, we can avoid auto layout complications when self-sizing.
+    switch indexPath.row {
+    case 0: // Header
+      return headerCell(for: indexPath.section)
+    case 1..<rowCount-1: // Schedule
+      let adjustedIndexPath = IndexPath(item: indexPath.item - 1, section: indexPath.section)
+      return scheduleCell(for: adjustedIndexPath)
+    case rowCount-1: // Footer
+      return footerCell(for: indexPath.section)
+    default:
+      fatalError("No row available.")
+    }
+  }
+
+  private func scheduleCell(for indexPath: IndexPath) -> UITableViewCell {
     let activity = activities[indexPath.section][indexPath.row]
     let isConcurrent = schedule[indexPath.section].activities.first(where: { $0.id == activity.id }) == nil
 
     let activityCell: ScheduleCell
     if isConcurrent {
-      guard let cell = tableView.dequeueReusableCell(withIdentifier: "ConcurrentCell") as? ConcurrentCell else { return UITableViewCell() }
+      guard let cell = tableView.dequeueReusableCell(withIdentifier: "ConcurrentCell") as? ConcurrentCell else {
+        fatalError("This should always return a cell.")
```

Comment:
Although this should never happen, Wouldn't it be better to be on the safe side and return `UITableViewCell()`, for the end user I think they might prefer a blank cell over an app crash, right?

Reply:
I'd argue otherwise; the 'should never happen' indicates that it should never be exposed even to the end user on accident. With a hard crash, you can be pretty sure you'll catch this during development and properly fix it. With a blank cell, you might just think `¯\_(ツ)_/¯` or not even notice it.

I can change this for an `assertFailure(...); return UITableViewCell()`, since that has the same effect, but for actual programming errors, I still prefer `fatalError(...)`.

Let me know.

Reply:
Sound like a great idea; best of both worlds.